pkg/ibmvpc/subnetsIPBlocks.go

```diff
@@ -0,0 +1,237 @@
+	for _, subnetObj := range rc.SubnetList {
+		subnetsBlocks[*subnetObj.CRN].splitByFiltersBlocks =
+			splitSubnetOriginalBlock(subnetsBlocks[*subnetObj.CRN].subnetOriginalBlock, filtersBlocks[*subnetObj.VPC.CRN])
+	}
+}
+
+func splitSubnetOriginalBlock(subnetOriginalBlock *ipblock.IPBlock, filtersBlocks []*ipblock.IPBlock) []*ipblock.IPBlock {
+	filtersBlocksOnSubnet := []*ipblock.IPBlock{}
+	for _, filterBlock := range filtersBlocks {
+		filterBlocksOnSubnet := subnetOriginalBlock.Intersect(filterBlock)
+		if !filterBlocksOnSubnet.IsEmpty() {
+			filtersBlocksOnSubnet = append(filtersBlocksOnSubnet, filterBlocksOnSubnet)
+		}
+	}
+	return filtersBlocksOnSubnet
+}
+
+// getSubnetsFreeBlocks() - calc all the addresses that are not allocated:
+// for each subnet:
+//  1. make a copy of the splitByFiltersBlocks
+//  2. remove the addresses that was already allocated
+func (subnetsBlocks subnetsIPBlocks) getSubnetsFreeBlocks(rc *datamodel.ResourcesContainerModel) error {
+	for _, subnetObj := range rc.SubnetList {
+		subnetsBlocks[*subnetObj.CRN].freeAddressesBlocks = make([]*ipblock.IPBlock, len(subnetsBlocks[*subnetObj.CRN].splitByFiltersBlocks))
+		for i, b := range subnetsBlocks[*subnetObj.CRN].splitByFiltersBlocks {
+			subnetsBlocks[*subnetObj.CRN].freeAddressesBlocks[i] = b.Copy()
+		}
+		// all the allocated IPs are at subnetObj.ReservedIps.
+		for blockIndex := range subnetsBlocks[*subnetObj.CRN].splitByFiltersBlocks {
+			for _, reservedIP := range subnetObj.ReservedIps {
+				if err := subnetsBlocks.removeAddressFromFree(*reservedIP.Address, *subnetObj.CRN, blockIndex); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// allocSubnetFreeAddress() allocated a free address from a block (for the private ip):
+func (subnetsBlocks subnetsIPBlocks) allocSubnetFreeAddress(subnetCRN string, blockIndex int) (string, error) {
+	address := subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex].FirstIPAddress()
+	return address, subnetsBlocks.removeAddressFromFree(address, subnetCRN, blockIndex)
+}
+func (subnetsBlocks subnetsIPBlocks) subnetBlocks(subnetCRN string) []*ipblock.IPBlock {
+	return subnetsBlocks[subnetCRN].splitByFiltersBlocks
+}
+
+func (subnetsBlocks subnetsIPBlocks) removeAddressFromFree(address, subnetCRN string, blockIndex int) error {
+	addressBlock, err := ipblock.FromIPAddress(address)
+	if err != nil {
+		return err
+	}
+	subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex] =
+		subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex].Subtract(addressBlock)
+	return nil
+}
+
+// ///////////////////////////////////////////
+// filtersBlocks create a slice of disjoint blocks, split according to the rules blocks, and their sum is allCidr:
+// 1. collect a slice all the acl blocks
+// 2. add to the slice all the gs blocks
+// 3. add to the slice the CidrAll block
+// 3. create a list of disjoint blocks from this slice
+type filtersBlocks map[string][]*ipblock.IPBlock
```

Comment:
type documentation should describe what this type holds/represents, and not a series of actions.. these actions should be moved to the relevant functions, such as `getFiltersBlocks`.
The type documentation can be modified such as the following:
`filtersBlocks is a map from VPC UID to a list of IPBlocks, which holds the list of disjoint IPBlocks computed from all referenced CIDRs in that VPC's filters rules`

Reply:
fixed